src/PlatformNames.jl

```diff
@@ -0,0 +1,110 @@
+export supported_platforms, platform_triplet, platform_key
+
+const platform_to_triplet_mapping = Dict(
+    :linux64 => "x86_64-linux-gnu",
```

Comment:
there are many 64 bit linux platforms. instead of inventing new naming conventions, why not use standardized gnu triplets directly?

Reply:
This is to normalize the small, unimportant variations between multiple triplets that we don't care about.  E.g. we don't care about the difference between `x86_64-linux-gnu` and `x86_64-pc-linux-gnu` or `x86_64-unknown-linux-gnu`.  The whole point of this package is to make a single binary that can run on any `x86_64` machine that is running on a `linux` kernel.  And as a side note, it's easier to type `:linux64` than to remember the whole triplet.

Reply:
These won't run on any x86_64 machine with a linux kernel, they'll also require a sufficiently recent glibc. Python wheels call this "manylinux" but they're a bit more specific about what else that requires. I'd normalize to a single gnu triple per actual platform, whichever the build environment's recommended compilers use.

Who will need to type these? The list of filenames and urls can be output for copy-pasting when the build process is done. It's less regular for scripts to have these match the standard architecture name some of the time, but shortened to just bitness for x86.

Reply:
> The list of filenames and urls can be output for copy-pasting when the build process is done.

Indeed, I agree, but there are still cases where people type this, such as when calling `runshell(:linux64)` to open up a shell within the build environment for that particular target.  Yes, it's a small thing, but it's still useful, IMO.

> These won't run on any x86_64 machine with a linux kernel, they'll also require a sufficiently recent glibc.

That's not something that is encoded within the triplet, so I don't see that as a point in favor either way.

Reply:
That's what the gnu means - not in terms of version, but glibc at least. Anything with a different libc would not have gnu in its triplet.

Reply:
> Anything with a different libc would not have gnu in its triplet.

That's a good point; Let's cross that bridge when we come to it.  Just as I am focusing only on platforms for which we have "official" Julia binary support right now (well, in actuality, it's whatever I have personally built Julia binaries for), I'm going to keep the scope limited for now.  We can always add to this shorthand list in the future, I think that's actually an argument for maintaining a shorthand list, as it allows us to map from the large set of possible input triplets to a manageable shortlist.

Reply:
Bridge already crossed, there's an alpine linux package of julia and it shouldn't be downloading glibc binaries.

Some normalization is needed, but there's valuable information you're dropping if you try to use "linux64" as a name for anything, that's too ambiguous. Better to normalize to an equivalent triple instead of making up now conventions.

Reply:
> Bridge already crossed, there's an alpine linux package of julia and it shouldn't be downloading glibc binaries.

There are lots of packages of Julia for exotic systems, and for all of these systems, I intend to add support as usage increases to the point of it being a worthwhile use of my time.  As it stands, adding support for non-glibc Linux distros is lower on my list than ensuring things like armv7l and aarch64 support work.

That being said, adding support is easy, we make a separate normalization such as `:linux64musl` or something.  I am confident enough in `glibc`'s dominance to consider `:linux64` to be a reasonable shorthand for the most common 64-bit linux configuration.

> Better to normalize to an equivalent triple instead of making up new conventions.
I disagree, calling the literal `x86_64-redhat-linux` the same thing as the literal `x86_64-suse-linux` is more confusing than coming up with our own, semi-arbitrary shorthand that everything is mapped to.  Even though there exists a reasonable middle ground in this case (`x86_64-unknown-linux`) I would much rather treat triplets as an implementation detail and have this software do the reasonable thing for 99% of all Julia users.

In the end, this is all just naming anyway, there is no technical problem here, and as you stated earlier, most of this is going to be copy-pasted from autogenerated output anyway, so I don't see the harm of designing for humans a little bit more and not using triplets that are needlessly fragmented for our use cases.

Reply:
Triplets already exist and are encoded in the build information, versioninfo output, etc. Your new naming shorthand does not. I'm not suggesting adding support for non-glibc linux distros immediately, I'm suggesting not inventing a naming convention that is misleading and wrong for them. They already have their own triple.

Reply:
The problem is that we want some of the information of the triplet, not all of it.  E.g. we don't care about the middle field of the triplet, we care about the first and third elements.  There's no reason to use a triplet to denote something that triplets were not intended to denote.  Using a single triplet to refer to `x86_64-linux` and `x86_64-redhat-linux` and x86_64-unknown-linux-gnu` is wrong and misleading, but referring to them all as `:linux64` seems much more straightforward and honest to me.

From an API standpoint, I also think using a `Symbol` to represent the platform gets across the idea that there is a small, finite, canonical set of choices, whereas a `String` doesn't quite convey that information as well.  This is an aesthetic choice, with not much weight in it, but it's another reason why I think we should keep this "arbitrary mapping".

Reply:
It's not using a single triplet to refer to multiple - it's building for a single canonical triple, and encoding the fact that a set of similar triples will be compatible and should use the same binaries. This is what triples were meant for, particularly considering all cross compilation infrastructure in autotools, gcc, and binutils works directly in terms of them. The vendor may be superfluous for our purposes, but we won't be building binaries with multiple vendor toolchain settings, just one. The kernel doesn't capture everything we need as the OS/libc information makes an important difference for compatibility, and there's no need to abbreviate the architecture as that just makes scripting less convenient with special cases. And who says we won't all be using 64 bit arm workstations in 5 or 10 years? Or one of the bigger distributions switches to musl at some point? Following the existing well-established convention directly is more future proof than making up our own.